File: src/DFA.js
define(['./Class', './Struct', './FARule', './Utils'], function(Class, Struct, FARule, Utils){

  eval(Utils.importScope('Utils'));

  var DFARulebook = Struct('DFARulebook', ['rules'])
    .method('next_state', function(state, character){
      return this.rule_for(state, character).follow();
    })
    .method('rule_for', function(state, character){
      return this.rules.reduce(function(last, rule){
        return last || (rule.is_applies_to(state, character) && rule || null )
      }, null);
    });

  var DFA = Struct('DFA', ['current_state', 'accept_states', 'rulebook'])
    .method('is_accepting', function(){
      var _this = this;
      return this.accept_states.reduce(function(last, state){
        return last || equal(_this.current_state, state);
      }, false);
    })
    .method('read_character', function(character){
      this.current_state = this.rulebook.next_state(this.current_state, character);
      return this;
    })
    .method('read_string', function(string){
      var _this = this;
      sliceStr(string).forEach(function(character){
        _this.read_character(character);
      });
      return this;
    });

  var DFADesign = Struct('DFADesign', ['start_state', 'accept_states', 'rulebook'])
    .method('to_dfa', function(){
      return DFA(this.start_state, this.accept_states, this.rulebook);
    })
    .method('is_accepts', function(string){
      return this.to_dfa().read_string(string).is_accepting();
    });

  return {
    FARule: FARule,
    DFARulebook: DFARulebook,
    DFADesign: DFADesign,
    DFA: DFA,
  }
});

File: Haskell/Automaton/DFA.hs

module DFA where

import Data.List
import FARule

data DFA s c = DFA s [s] (Rulebook s c)
    deriving(Eq, Show, Read)

nextState :: (Eq s, Eq c) => Rulebook s c -> s -> Input c -> s
nextState rb s c = follow $ ruleFor rb s c

ruleFor :: (Eq s, Eq c) => Rulebook s c -> s -> Input c -> Rule s c
ruleFor (Rulebook rs) s c = helper $ find (\r->isApplies r s c) rs
    where helper (Just n) = n
          helper (Nothing) = error "unexpected input"

accepting :: (Eq s) => DFA s c -> Bool
accepting (DFA current_state accept_states _) =
    any (== current_state) accept_states

readCharacter :: (Eq s, Eq c) => DFA s c -> Input c -> DFA s c
readCharacter (DFA current_state accept_states rulebook) character =
    DFA (nextState rulebook current_state character) accept_states rulebook

readString :: (Eq s, Eq c) => DFA s c -> [Input c] -> DFA s c
readString dfa (c:cs) = readString (readCharacter dfa c) cs
readString dfa [] = dfa

data DFADesign s c = DFADesign s [s] (Rulebook s c)
    deriving(Eq, Show, Read)

toDfa :: (Eq s, Eq c) => DFADesign s c -> DFA s c
toDfa (DFADesign start_state accept_states rulebook) = 
    DFA start_state accept_states rulebook

accepts ::(Eq s, Eq c) => DFADesign s c -> [c] -> Bool 
accepts dfa_design inputs =  accepting $
    readString (toDfa dfa_design) . map Input $ inputs
